pkg/services/ngalert/notifier/alertmanager.go

```diff
@@ -32,70 +32,141 @@ type Alertmanager struct {
 	notificationLog *nflog.Log
 	// silences keeps the track of which notifications we should not fire due to user configuration.
 	silences   *silence.Silences
-	marker     types.Marker
 	alerts     *AlertProvider
 	dispatcher *dispatch.Dispatcher
 
+	configReloadC chan struct{}
+
 	wg sync.WaitGroup
 }
 
 func init() {
-	registry.RegisterService(&Alertmanager{})
+	registry.RegisterService(&Alertmanager{
+		configReloadC: make(chan struct{}, 1),
+	})
 }
 
 func (am *Alertmanager) IsDisabled() bool {
 	return !setting.AlertingEnabled || !setting.ExecuteAlerts
 }
 
-func (am *Alertmanager) Init() error {
+func (am *Alertmanager) Init() (err error) {
 	am.logger = log.New("alertmanager")
+	am.notificationLog, err = nflog.New(
+		nflog.WithRetention(time.Hour*24),                         //TODO: This is a setting.
+		nflog.WithSnapshot(filepath.Join("dir", "notifications")), //TODO: This should be a setting
+	)
+	if err != nil {
+		return errors.Wrap(err, "unable to initialize the notification log component of alerting")
+	}
 
 	return nil
 }
 
 func (am *Alertmanager) Run(ctx context.Context) error {
+	reloadFromConfig := func() error {
+		cfg1, cfg2, err := getConfigFromDatabase()
+		if err != nil {
+			return errors.Wrap(err, "get config from database")
+		}
+		return errors.Wrap(am.reloadFromConfig(cfg1, cfg2), "reload config")
+	}
+	// Make sure dispatcher starts. We can tolerate future reload failures.
+	if err := reloadFromConfig(); err != nil {
+		return err
+	}
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				if am.dispatcher != nil {
```

Comment:
Can we have a `Stop` function that wraps together what we need to stop the alertmanager.